gateway/mw_streaming.go

```diff
@@ -0,0 +1,383 @@
+			err = json.Unmarshal([]byte(replacedStream), &unmarshaledStream)
+			if err != nil {
+				s.Logger().Errorf("Failed to unmarshal replaced stream config: %v", err)
+				continue
+			}
+			stream = unmarshaledStream
+		} else {
+			s.Logger().Debugf("No request available to replace variables in stream config for %s", streamID)
+		}
+	}
+	return streamsConfig
+}
+
+// createStream creates a new stream
+func (sm *StreamManager) createStream(streamID string, config map[string]interface{}) error {
+	streamFullID := fmt.Sprintf("%s_%s", sm.mw.Spec.APIID, streamID)
+	sm.mw.Logger().Debugf("Creating stream: %s", streamFullID)
+
+	stream := streaming.NewStream(sm.mw.allowedUnsafe)
+	err := stream.Start(config, &handleFuncAdapter{
+		mw:       sm.mw,
+		streamID: streamFullID,
+		muxer:    sm.muxer,
+		sm:       sm,
+		// child logger is necessary to prevent race condition
+		logger: sm.mw.Logger().WithField("stream", streamFullID),
+	})
+	if err != nil {
+		sm.mw.Logger().Errorf("Failed to start stream %s: %v", streamFullID, err)
+		return err
+	}
+
+	sm.streams.Store(streamFullID, stream)
+	sm.mw.Logger().Infof("Successfully created stream: %s", streamFullID)
+
+	return nil
+}
+
+func (sm *StreamManager) hasPath(path string) bool {
+	for _, p := range sm.listenPaths {
+		if strings.TrimPrefix(path, "/") == strings.TrimPrefix(p, "/") {
+			return true
+		}
+	}
+	return false
+}
+
+// ProcessRequest will handle the streaming functionality
+func (s *StreamingMiddleware) ProcessRequest(w http.ResponseWriter, r *http.Request, _ interface{}) (error, int) {
+	strippedPath := s.Spec.StripListenPath(r.URL.Path)
+	if !s.defaultStreamManager.hasPath(strippedPath) {
+		return nil, http.StatusOK
+	}
+
+	s.Logger().Debugf("Processing request: %s, %s", r.URL.Path, strippedPath)
+
+	newRequest := r.Clone(r.Context())
+	newRequest.URL.Path = strippedPath
+
+	var match mux.RouteMatch
+	streamManager := s.createStreamManager(r)
+	streamManager.routeLock.Lock()
+	streamManager.muxer.Match(newRequest, &match)
+	streamManager.routeLock.Unlock()
+
+	// direct Bento handler
+	handler, _ := match.Handler.(http.HandlerFunc)
+
+	handler.ServeHTTP(w, r)
+
+	return nil, mwStatusRespond
+}
+
+func (s *StreamingMiddleware) Unload() {
+	s.Logger().Debugf("Unloading streaming middleware %s", s.Spec.Name)
+
+	totalStreams := 0
+	s.streamManagers.Range(func(_, value interface{}) bool {
+		manager := value.(*StreamManager)
+		manager.streams.Range(func(_, _ interface{}) bool {
+			totalStreams++
+			return true
+		})
+		return true
+	})
+	globalStreamCounter.Add(-int64(totalStreams))
+
+	s.cancel()
+
+	s.Logger().Debug("Closing active streams")
+	s.streamManagers.Range(func(_, value interface{}) bool {
+		manager := value.(*StreamManager)
+		manager.streams.Range(func(_, streamValue interface{}) bool {
+			if stream, ok := streamValue.(*streaming.Stream); ok {
+				stream.Reset()
+			}
+			return true
+		})
+		return true
+	})
+
+	s.streamManagers = sync.Map{}
```

Comment:
Could this lead to a data race?

Reply:
I don't think so, it is called when the middleware is being unloaed i believe, plus the race condition check didn't seem to flag it.